design/ceph-nfs-ganesha.md

```diff
@@ -0,0 +1,190 @@
+# Ceph NFS-Ganesha CRD
+
+[NFS-Ganesha] is a user space NFS server that is well integrated with [CephFS]
+and [RGW] backends. It can export Ceph's filesystem namespaces and Object
+gateway namespaces over NFSv4 protocol.
+
+Rook already orchestrates Ceph filesytem and Object store (or RGW) on
+Kubernetes (k8s). It can be extended to orchestrate NFS-Ganesha server daemons
+as highly available and  scalable NFS gateway pods to the Ceph filesytem and
+Object Store. This will allow NFS client applications to use the Ceph filesytem
+and object store setup by rook.
+
+This feature mainly differs from the feature to add NFS as an another
+storage backend for rook (the general NFS solution) in the following ways:
+
+* It will use the rook's Ceph operator and not a separate NFS operator to
+  deploy the NFS server pods.
+
+* The NFS server pods will be directly configured with CephFS or RGW
+  backend setup by rook, and will not require CephFS or RGW to be mounted
+  in the NFS server pod with a PVC.
+
+## Design of Ceph NFS-Ganesha CRD
+
+The NFS-Ganesha server and its exports' settings will be exposed to rook as a
+Custom Resource Definition (CRD). Creating the nfs-ganesha CRD will launch
+NFS Ganesha server pods that will be configured with a rook CephFS filesystem
+or a RGW object store backend. The servers' exports will be stored in
+[k8s ConfigMaps]; the NFS client recovery data will be stored in a Ceph RADOS
+pool; and the servers will have stable IP addresses by using [k8s Service].
+This allows the servers pods to be scalable and highly available.
+
+### Prerequisities
+
+- A running rook Ceph filesystem or object store, whose namespaces will be
+  exported by the Ganesha servers.
+  e.g.,
+  ```
+  kubectl create -f filesystem.yaml
+  ```
+
+- An existing RADOS pool (e.g., CephFS's data pool) or a pool created with a
+  [Ceph Pool CRD] to store NFS client recovery data.
+
+
+### Ceph NFS-Ganesha CRD
+
+The NFS-Ganesha CRD will specify the following:
+
+- Number of active Ganesha servers
+
+- Placement of the Ganesha servers
+
+- Resource limits (memory, CPU) of the Ganesha pods
+
+- Export settings
+
+- RADOS pool and namespace where the NFS client recovery data will be stored
+
+
+Below is an example NFS-Ganesha CRD, `nfs-ganesha.yaml`
+
+```yaml
+apiVersion: ceph.rook.io/v1alpha1
+kind: NFSGanesha
+metadata:
+  # The name of Ganesha server cluster to create, which will be reflected in
+  # the name(s) of the ganesha server pod(s)
+  name: mynfs
+  # The namespace of the Rook cluster where the Ganesha server cluster is
+  # created.
+  namespace: rook-ceph
+spec:
+  # The Ceph filesystem or RGW that will be configured as the Ganesha server's
+  # backend. It will be the identified by the name of the filesytem or object
+  # store CRD.
+  backendName: myfs
```

Comment:
how do we know if it's a filesystem or object store? What if we break this into two settings?
```
  store:
    type: [object | file]
    name: myfs
```